DM-49622: add Pydantic serialization for DimensionRecordSet

TallJimbo opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

TallJimbo committed:
Fix copy/paste bugs in arrow utils.

TallJimbo committed:
Typing fixes from using pyarrow-stubs.

I tried the third-party annotations from pyarrow-stubs locally, and
while that helped me spot a few unambiguous fixes (on this commit), it
also generated a lot of false positives due to the stubs being a out of
date and/or just inaccurate, as well as a lot of complaints that I
couldn't easily verify.

TallJimbo committed:
Stop permitting None as a data ID value (in a few places).

None hasn't ever been a valid data ID value in the vast majority, but
in one important type alias and a bit of logic in expandDataId (that
never fired in practice), we were still worried about it.

TallJimbo committed:
Add annotated type alias for serializing bytes as hex.

TallJimbo committed:
Fix ColumnSpec for bytes by using hex-based annotation.

Previous version assumed Pydantic knew what to do with 'bytes'
natively; it doesn't.

TallJimbo committed:
Always use type adapter in ColumnSpec pydantic conversions.

This may be a bit slower, but it doesn't leave a validation gap.

TallJimbo committed:
Add pydantic_core utility functions to ColumnSpec system.

These should let us use Pydantic's low-level Rust-implemented
validation and serialization for mappings and tuples that represent
serialized dimension records and query rows, where the types are not
known statically, but are known up front.

TallJimbo committed:
Use pydantic_core validator/serializer for general query results.

TallJimbo committed:
Handle nullable fields in ColumnSpec pydantic_core schema.

TallJimbo committed:
Mark dataset timespan columns in queries as nullable.

These are always null when querying non-calibration collections,
even if they are never null in calibration collections.

TallJimbo committed:
Add container-friendly serialization to DimensionRecord.

TallJimbo committed:
Don't shortcut on data ID in DimensionRecord equality.

This confused me in tests, and it's bound to confuse users (especially
any looking to why their ingest or define-visits run failed with
conflicts).  The optimization just isn't worth it (though it probably
is in DimensionRecordSet, which still only uses data IDs for
comparisons).

TallJimbo committed:
Add serialization to DimensionRecordSet.

TallJimbo committed:
Add changelog entry.

TallJimbo commented on python/lsst/daf/butler/dimensions/_records.py:
```diff
@@ -490,6 +538,86 @@ def toDict(self, splitTimespan: bool = False) -> dict[str, Any]:
 
+    def serialize_key_value(self) -> SerializedKeyValueDimensionRecord:
+        """Serialize this record to a `list` that can be sliced into a key
+        (data ID values) / value (everything else) pair.
```
If anybody has thoughts on better nomenclature for this new `DimensionRecord` serialization mode (it needs to be distinguished from the old one somehow), I'd love to hear it, because I'm not thrilled with what I've come up with so far.

timj replied:
It's still a single DimensionRecord? not effectively `list[DimensionRecord]`?

TallJimbo replied:
Yes, here.  Serializing a collection of dimension records is over on `DimensionRecordSet.serialize_records` and it delegates to this.

dhirving commented on python/lsst/daf/butler/column_spec.py:
```diff
@@ -243,6 +250,157 @@ def display(self, level: int = 0, tab: str = "  ") -> list[str]:
+            )
+            for spec in columns
+        }
+    )
```
There's a functional form of TypedDict that could accomplish this without resorting to pydantic-core, e.g. [from the docs](https://docs.python.org/3/library/typing.html#typing.TypedDict)
```
Point2D = TypedDict('Point2D', {'x': int, 'y': int, 'label': str})
```

TallJimbo replied:
Done (for `tuple` as well).  I considered this before but didn't really give it a try because I'm uncomfortable using the static-typing constructs as runtime objects.  But I think I agree that in this case it's a bit better when you factor in the vague stability guarantees on `pydantic_core`.

dhirving commented on python/lsst/daf/butler/column_spec.py:
```diff
@@ -243,6 +250,157 @@ def display(self, level: int = 0, tab: str = "  ") -> list[str]:
+        A low-level Pydantic schema that can be used to construct a validator,
+        serializer or JSON schema.
+    """
+    return pydantic_core.core_schema.tuple_schema([spec.pydantic_core_schema for spec in columns])
```
You can create tuple types dynamically without pydantic-core as well:
```
types = [int, str, str]
tuple[*types]

dhirving reviewed (approved):
Looks good to me.  Not sure it's necessary to resort to pydantic-core here but it's also used in a relatively straightforward way so it's not necessarily a problem either.  

The changes to general query result serialization are fine with me -- nothing is really using that except for Stelio's SIA service, which is not widely used yet, so if it breaks something we'll fix it.

TallJimbo committed:
Use static-typing constructs at runtime instead of pydantic_core.

TallJimbo committed:
Remove ColumnSpec make_dict_type_adapter (for now, at least).

This requires Python 3.12+ or typing_extensions.TypedDict, and since
we don't need this function now it's easier to drop it.  This commit
can be reverted if we want it back.
